Use modern __atomic builtins in atomic.c

andreabedini opened the pull request:
Replace legacy __sync_fetch_and_* builtins with their modern __atomic_fetch_* equivalents. This simplifies the code significantly, particularly for the nand operation which previously required extensive workarounds for compiler compatibility issues.

Changes:

- Replace __sync_fetch_and_{add,sub,and,or,xor} with __atomic_fetch_*

- Replace __sync_fetch_and_nand with __atomic_fetch_nand

- Remove CAS-based fallback for nand operations

- Remove compiler-specific warning suppressions for -Wsync-nand

- Remove volatile qualifiers (not needed with __atomic builtins)

- Update comments to reflect modern atomics usage

All operations maintain __ATOMIC_SEQ_CST memory ordering for sequential consistency, matching the original behavior.

hsyl20 commented on libraries/ghc-internal/cbits/atomic.c:
```diff
@@ -76,171 +76,90 @@ extern StgWord hs_atomic_and8(StgWord x, StgWord val);
-#pragma GCC diagnostic ignored "-Wsync-nand"
-#endif
+// The modern __atomic_fetch_nand has well-defined semantics:
+// it atomically replaces *ptr with ~(*ptr & val) and returns the old value.
```
Isn't the original note bogus? I think it has the semantics change reversed, no?

andreabedini replied:
Likely bogus. I removed the comment.

hsyl20 reviewed (approved):
LGTM. We should upstream this if it works on all supported platforms.

I was put off by the Note [__sync_fetch_and_nand usage] but I think the note is wrong (see inline comment).

hasufell commented:
It was moved to `rts/prim/atomic.c` recently: https://gitlab.haskell.org/ghc/ghc/-/blob/master/rts/prim/atomic.c

angerman committed:
Use modern __atomic builtins in atomic.c

Replace legacy __sync_fetch_and_* builtins with their modern __atomic_fetch_*
equivalents. This simplifies the code significantly, particularly for the nand
operation which previously required extensive workarounds for compiler
compatibility issues.

Changes:
- Replace __sync_fetch_and_{add,sub,and,or,xor} with __atomic_fetch_*
- Replace __sync_fetch_and_nand with __atomic_fetch_nand
- Remove CAS-based fallback for nand operations
- Remove compiler-specific warning suppressions for -Wsync-nand
- Remove volatile qualifiers (not needed with __atomic builtins)
- Update comments to reflect modern atomics usage

All operations maintain __ATOMIC_SEQ_CST memory ordering for sequential
consistency, matching the original behavior.

Co-authored-by: Andrea Bedini <andrea@andreabedini.com>
